microsoft-edge/webview2/gettingstarted/winUI.md

```diff
@@ -0,0 +1,293 @@
+        <Button x:Name="myButton" Grid.Column="1" Click="myButton_Click">Go</Button>
+
+        <controls:WebView2 x:Name="MyWebView"  controls:Grid.Row="1" controls:Grid.ColumnSpan="2" 
+        Source="https://www.microsoft.com" HorizontalAlignment="Stretch" VerticalAlignment="Stretch" />
+
+    </Grid>
+    ```  
+    
+
+1.  In **MainWindow.xaml.cs**, copy the following code snippet to edit the `myButton_Click` method, which navigates the WebView to the URL entered in the address bar.  
+    
+    ```csharp
+    private void myButton_Click(object sender, RoutedEventArgs e)
+    {
+        try
+        {
+            Uri targetUri = new Uri(addressBar.Text);
+            MyWebView.Source = targetUri;
+        }
+        catch (FormatException ex)
+        {
+            // Bad address.
+        }
+    }
+    ```  
+    
+    Press `F5` to build and run your project.  Enter a new URL in the address bar, and select **Go**.  For example, enter `https://www.bing.com`.  Confirm that the WebView2 control navigates to the URL.  
+    
+    > [!NOTE]
+    > Make sure a complete URL is entered in the address bar.  An `ArgumentException` is thrown if the URL does not start with `http://` or `https://`.  
+    
+    :::image type="complex" source="./media/winui-gettingstarted-bing.png" alt-text="Bing":::
+       Bing
+    :::image-end:::
+    
+## Step 5 - Navigation events  
+
+The application that hosts WebView2 controls listens to the following events that are raised by the WebView2 control during navigation to web pages.  
+
+*   `NavigationStarting`  
+*   `SourceChanged`  
+*   `ContentLoading`  
+*   `HistoryChanged`  
+*   `NavigationCompleted`  
+
+For more information, see [Navigation Events](../reference/win32/0-9-488/icorewebview2.md#navigation-events).  
+ 
+When an error occurs, the following events are raised and may depend on navigation to an error page.  
+
+*   `SourceChanged`  
+*   `ContentLoading`  
+*   `HistoryChanged`  
+
+When there is an HTTP redirect, there are multiple `NavigationStarting` events.  
+
+To demonstrate how to use these events, start by registering a handler for `NavigationStarting` that cancels any requests that do not use HTTPS.  
+
+In `MainWindow.xaml.cs`, modify the constructor as shown below and add the `EnsureHttps` function.  
+
+```csharp
+public MainWindow()
+{
+    InitializeComponent();
+    webView.NavigationStarting += EnsureHttps;
+}
+
+private void EnsureHttps(WebView2 sender, WebView2NavigationStartingEventArgs args)
+{
+    String uri = args.Uri;
+    if (!uri.StartsWith("https://"))
+    {
+        args.Cancel = true;
+    }
+    else
+    {
+        addressBar.Text = uri;
+    }
+}
+```  
+
+In the constructor, EnsureHttps is registered as the event handler on the `NavigationStarting` event on the WebView2 control.  
+
+Press `F5` to build and run your project.  Confirm that when navigating to an HTTP site, the WebView **remains unchanged**.  However, the WebView navigates to HTTPS sites.  
```

Comment:
Note:  If I try to navigate to "http://bing.com", the navigation gets cancelled, but the NavigationStarting event fires multiple times, which results in the "else" block updating the addressBar even though navigation didn't happen.  Is that a bug in core WebView2 that it doesn't block further navigation events?

Here is what I see with some debug output on when the Source property is set, when EnsureHttps gets called, and whether the Cancel line ran:

Setting Source to: http://bing.com/
Navigating to: http://bing.com/
   Cancelling!
Navigating to: https://bing.com/
Navigating to: https://www.bing.com/?toWww=1&redig=34F674EDB5864A4B9C8ABEB69DFA9F92